### ISSUE TITLE
mirror `dataclasses.replace` behaviour?

### ISSUE THREAD

**Issue by epignatelli** [patrick-kidger/equinox#426]
Hi Patrick, thanks for `equinox` -- it works great and easy to jump in.
I found some patters very clear, but harder to jump in without prior knowledge.

`equinox.tree_at` is one of these.
Would you consider mirroring the canonical `dataclass` interface?

For example, 
```python
    def replace(self, **kwargs) -> ReplayBuffer:
        values = [kwargs[k] for k in kwargs]
        return eqn.tree_at(lambda x: [getattr(x, k) for k in kwargs], self, values)
```

I am sure there can be something more efficient than this, but it is enough to explain the idea.

**Comment by patrick-kidger** [comment]
Right! Whilst the `lambda` approach might be a completely general way to navigate pytrees, in practice 99% of use-cases just involve navigating via item-getting and attr-getting, and folks seem to find the `lambda` approach a bit confusing.

I'm wondering about generalising what you've written to something like:
```python
class tree_modify:
    def __init__(self, tree, path=()):
        self.tree = tree
        self.path = path

    def __getitem__(self, item):
        return tree_modify(self.tree, self.path + ("getitem", item)))

    def __getattr__(self, item):
        return tree_modify(self.tree, self.path + ("getattr", item))

    def __lshift__(self, value):
        def _get(tree):
            for kind, item in self.path:
                if kind == "getitem":
                    tree = tree[item]
                elif kind == "getattr":
                    tree = getattr(tree, item)
                else:
                    assert False
            return tree
        return eqx.tree_at(_get, self.tree, value)
```
which you can then use as:
```python
mlp = eqx.nn.MLP(...)
new_linear = eqx.nn.Linear(...)
new_mlp = eqx.tree_modify(mlp).layers[-1] << new_linear
```
Although maybe the `<<` syntax is too magic?

Some kind of nicer interface seems to make sense, anyway.

**Comment by epignatelli** [comment]
I actually like the `lambda` approach, once I got how it works!
It is just takes longer to get than replace does (to me at least), and since `equinox` strength is a low learning curve, I think `replace` might help.

Yeah, I think using `lshift` might be not super intuitive either.
What about resorting to jax/python indexing to generalise the above?

```python
import equinox as eqx
from jax import Array
import jax.numpy as jnp


class Foo(eqx.Module):
  bar: Array

  def replace(self, **kwargs):
      values = [kwargs[k] for k in kwargs]
      return eqn.tree_at(lambda x: [getattr(x, k) for k in kwargs], self, values)

bar = jnp.zeros((2,))
foo = Foo(bar=bar)

baz = jnp.asarray(1)
baz = bar.at[0].set(baz)
foo = foo.replace(bar=baz)
print(foo.bar)
# Array([1., 0.], dtype=float32)
```

We can either leave how select the element to the user or generalise `replace` to use jax vs python indexing according to whether the field is dynamic or static.

**Comment by epignatelli** [comment]
Another idea might be to mirror and extend jax indexing helper to `pytree`s.

E.g.,
```python
mlp.at[mlp.layers].set(new_linear)
```
or even, because I don't like the `mlp` repetition,

```python
mlp.at['layers'].set(new_linear)
```

This might even reuse `eqx.tree_at` with:

```python
class Module:
  @property
  def at(self):
    return eqx.IndexingHelper
```

**Comment by ASEM000** [comment]
Check [here](https://github.com/ASEM000/PyTreeClass) for generalized concept based on lenses, and [here](https://pytreeclass.readthedocs.io/en/latest/notebooks/common_recipes.html#%5B7%5D-Use-PyTreeClass-with-Flax/Equinox) on how to integrate with flax/equinox (or any registered pytree)

**Comment by patrick-kidger** [comment]
Haha, I actually remembered and took a look at your work @ASEM000 when I was writing this. IIUC, there's no way to distinguish between `x["foo"]` and `x.foo`: a string always refers to the latter, but this makes it impossible to look up values in a `dict[str, Any]`. I think @epignatelli's `mlp.at['layers', -1].set(new_linear)` has the same issue.

I would say the main two criteria are:
- this should be a free function, not a method. Right now `Module`s are never special-cased relative to any other pytree -- in particular, they have no extra methods.
- this should offer a general way to mutate nested leaves, not just at depth-1 like the `replace` function.

**Comment by ASEM000** [comment]
The example below should answer the questions, notably:
1) it can work on dict/class. the distinction between dict/class attribute is dealt with on the path entry level defined using jax registration, and accessed using `jax.tree_map_with_path`.
2) you don't need a method, you can use `pytc.AtIndexer(pytree).at[...]`. the .at property wraps `AtIndexer`.
3) allows nesting, 
4) it's not special cased to any pytree. as long as a pytree defined using jax key paths (like equinox module). you can check  [the source](https://github.com/ASEM000/PyTreeClass/blob/main/pytreeclass/_src/tree_index.py)
5) you can do much more, as outlined in the links above.

If you have any questions let me know.



```python

import pytreeclass as pytc
import equinox as eqx
from typing import Any

print(pytc.__version__)
# '0.3.8'


class Tree(eqx.Module):
    some_dict: dict[str, Any]
    some_list: list[int]


tree = Tree(some_dict={"a": 1, "b": [1, 2, 3]}, some_list=[1, 2, 3])

print(pytc.AtIndexer(tree).at["some_dict"].at["a"].get())
# Tree(some_dict={'a': 1, 'b': [None, None, None]}, some_list=[None, None, None])

print(pytc.AtIndexer(tree).at["some_dict"].at["b"].at[0].get())
# Tree(some_dict={'a': None, 'b': [1, None, None]}, some_list=[None, None, None])

```

**Comment by ASEM000** [comment]
> As a user, I think I'd prefer the lambda approach to the chain of at because it is more concise and direct.

You can try this if you do not prefer chained `.at`, `at` is just syntactic sugar.

```python

import pytreeclass as pytc
import equinox as eqx
from typing import Any

class Tree(eqx.Module):
    some_dict: dict[str, Any]
    some_list: list[int]


tree = Tree(some_dict={"a": 1, "b": [1, 2, 3]}, some_list=[1, 2, 3])

print(pytc.AtIndexer(tree, where=["some_dict", "a"]).get())
# Tree(some_dict={'a': 1, 'b': [None, None, None]}, some_list=[None, None, None])

print(pytc.AtIndexer(tree, where=["some_dict", "b", 0]).get())
# Tree(some_dict={'a': None, 'b': [1, None, None]}, some_list=[None, None, None])

```

- Among the objectives of  `PyTreeClass` is to combine oop+fp concepts for intuitive yet safe nested structure (pytrees) manipulation.  The `.at`/`AtIndexer` is similar to Haskell lenses which is a popular way to do such an operation functionally.

- I think a good example to compare the two approaches, is to try to replicate the example in the [readme here](https://github.com/ASEM000/PyTreeClass) using `tree_at`.


If you have any questions let me know.

**Comment by epignatelli** [comment]
Thanks for the example, @ASEM000.
We have different objectives here, though -- see my comment above.

**Comment by epignatelli** [comment]
Thanks @LouisDesdoigts !

I honestly think that the approach is `equinox` is the best approach for deep and complex updates.

This feature request had a different purpose: to expose the `dataclasses.replace` method for
A) a gentler transition to equinox (which is the main reason behind its philosophy, if I understand correctly), and
B) better syntax only for _simpler_ and shallower updates

The proposal was for the two approaches to _coexist_, not for one to replace the other

**Comment by patrick-kidger** [comment]
Right now I don't have any strong feelings either way! I generally prefer not to have simple-use-case-only versions of things, just because it helps upskill novice users into the more general way of doing things. But that's not a hard-and-fast rule.
